<a href="https://colab.research.google.com/github/Filgueiras/bootcamp_datascience/blob/main/Bootcamp_Desafio_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Projeto Módulo 02 - Alura - Data Science Aplicada**

###Autor

* **Marco Antonio Filgueiras Guimarães**
* Github: https://github.com/Filgueiras
* E-mail: marco.filgueiras@gmail.com / marco.filgueiras@hotmail.com
* LinkedIn: https://www.linkedin.com/in/marcofilgueiras/

### Sobre os dados (e a coleta)

Os dados do SUS vieram do [Data SUS - Imunização](http://tabnet.datasus.gov.br/cgi/tabcgi.exe?pni/cnv/cpniuf.def), conforme orientado no Discord.



Depois de explorar um pouco os resultados, vi que em total de doses ou em % de imunização, os dados de ano não fazem parte da tabela e que as vacinas mudaram ao longo dos anos.
Assim, resolvi fazer o seguinte:
1. Ver dados em números de doses.
2. Tem um arquivo para cada ano, desde 1994 até 2019 (ele sempre trazia números absolutos, não quebrando por ano, então tive que separar). Isso porque o sistema não nos mostra UF e ano no mesmo resultado (repetindo dados por ano ou algo assim). Ou faz uma coisa, ou outra...
> Ano a ano, ele coloca na primeira linha antes dos dados a informação de ano como "Período: AAAA" (ano com 4 dígitos).
3. Baixar os arquivos com linhas nas UFs e colunas nas vacinas ("imunos")
> Isso deixou o número de linhas do arquivo fixas (sempre igual a UFs) e as colunas variáveis (vacinas/imunos).
* Pq UF nas linhas: a importação inicial fica facilitada (por causa daquele monte de comentários que vem no final do arquivo, o skip no footer torna-se fixo). Eu não sei quantas vacinas/imunos há por ano, mas UF faz tempo que não muda (última UF é Tocantins, criada em 1988 a partir do norte de Goiás).
* Depois faço um MELT, lá no meio...
> [Tocantins na Wikipedia](https://en.wikipedia.org/wiki/Tocantins)

### **Meus objetivos**

>1. Verificar quais vacinas eram dadas ao longo dos anos.
>2. Verificar a desigualdade de distribuição das vacinas (se um estado recebe e outro não, no mesmo período).
>3. Verificar se alguma vacina deixou de ser dada e depois voltou ao calendário. Caso haja tempo, vou verificar os motivos.

###Formulário de envio

> Enviar para Alura em: https://forms.gle/HNR6yen2xB5ttv7D6

### **Início da Análise / Show me the Code**

>Importar todos os dados de vacina incluídos na minha pasta de projeto no Github (repara no skipfooter=19, para ignorar os comentários no final e a linha totalizadora, que me fez chegar a 20 no skip).

In [ ]:
import pandas as pd

#lista de arquivos a importar
lista_arquivos = ["A202931189_28_143_208-p2019.csv","A203022189_28_143_208-p2018.csv","A203103189_28_143_208-p2017.csv","A203140189_28_143_208-p2016.csv",
                 "A203214189_28_143_208-p2015.csv","A204149189_28_143_208-p2014.csv","A204349189_28_143_208-p2013.csv","A204440189_28_143_208-p2012.csv",
                 "A204553189_28_143_208-p2011.csv","A204031189_28_143_208-p2010.csv","A204658189_28_143_208-p2009.csv","A205022189_28_143_208-p2008.csv",
                 "A205129189_28_143_208-p2007.csv","A205245189_28_143_208-p2006.csv","A203934189_28_143_208-p2005.csv","A205332189_28_143_208-p2004.csv",
                 "A205409189_28_143_208-p2003.csv","A205537189_28_143_208-p2002.csv","A205613189_28_143_208-p2001.csv","A203838189_28_143_208-p2000.csv",
                 "A205700189_28_143_208-p1999.csv","A205745189_28_143_208-p1998.csv","A205820189_28_143_208-p1997.csv","A205849189_28_143_208-p1996.csv",
                 "A203730189_28_143_208-p1995.csv", "A203620189_28_143_208-p1994.csv"]


In [ ]:
def importa_arquivos_imuno_sus(caminho, nome, ano):
    dadosTemp = pd.read_csv(f"{caminho}{nome}",
                      encoding="ISO-8859-1",sep=";", skiprows=3, skipfooter=20, thousands=".",decimal=",", engine="python")
    dadosTemp["Ano"] = ano
    dadosTemp["UF"] = dadosTemp["Unidade da Federação"].str[2:]
    dadosTemp["UF"] = dadosTemp["UF"].str.strip()
    dadosTemp = dadosTemp.drop("Total", axis=1)
    dadosTemp = dadosTemp.drop("Unidade da Federação", axis=1)
    dadosTemp.set_index("UF","Ano")
    return dadosTemp

Importando todos os arquivos (poderia virar uma função, não virou, mas está pronto para isso)

In [ ]:
ordem = "d"
contador = 0
ano_atual = 2019

for nome in lista_arquivos:
  if (ano_atual - contador) == 2015:
    print("Pulando 2015 pro problemas no arquivo - formato errado")
  else:
    dataSetA = importa_arquivos_imuno_sus("https://raw.githubusercontent.com/Filgueiras/bootcamp_datascience/main/Desafio02/", lista_arquivos[contador], (ano_atual - contador))
    print(f"Importado com sucesso em {ano_atual - contador}")
    if contador == 0:
      dataSetSUS = dataSetA
    else:
      dataSetSUS = pd.concat([dataSetSUS,dataSetA])
  contador = contador + 1

Importado com sucesso em 2019
Importado com sucesso em 2018
Importado com sucesso em 2017
Importado com sucesso em 2016
Pulando 2015 pro problemas no arquivo - formato errado
Importado com sucesso em 2014
Importado com sucesso em 2013
Importado com sucesso em 2012
Importado com sucesso em 2011
Importado com sucesso em 2010
Importado com sucesso em 2009
Importado com sucesso em 2008
Importado com sucesso em 2007
Importado com sucesso em 2006
Importado com sucesso em 2005
Importado com sucesso em 2004
Importado com sucesso em 2003
Importado com sucesso em 2002
Importado com sucesso em 2001
Importado com sucesso em 2000
Importado com sucesso em 1999
Importado com sucesso em 1998
Importado com sucesso em 1997
Importado com sucesso em 1996
Importado com sucesso em 1995
Importado com sucesso em 1994


Conferência rápida: 26 tabelas (anos) com 27 linhas cada, deveria dar um dataset com 26*27 registros (resultado abaixo)

In [ ]:
display(len(lista_arquivos)*27)

702

Vamos ver quantas linhas tem o dataset unificado... (torcendo para ter 702)

In [ ]:
dataSetSUS.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 659 entries, 0 to 14
Data columns (total 27 columns):
 #   Column                                         Non-Null Count  Dtype  
---  ------                                         --------------  -----  
 0   072 BCG                                        659 non-null    int64  
 1   099 Hepatite B  em crianças até 30 dias        135 non-null    float64
 2   061 Rotavírus Humano                           351 non-null    float64
 3   053 Meningococo C                              243 non-null    float64
 4   073 Hepatite B                                 659 non-null    object 
 5   080 Penta                                      189 non-null    float64
 6   012 Pneumocócica                               243 non-null    float64
 7   074 Poliomielite                               659 non-null    object 
 8   100 Poliomielite 4 anos                        81 non-null     object 
 9   006 Febre Amarela                              659 non-

**Uma conferência mínima dos números exigiria verificar se há 27 linhas por ano...**
> Aqui é uma mapa geral... podemos ver as vacinas na linha do tempo e verificar a quantidade de UFs onde foram aplicadas. Qualquer número menor que 27 significa que não foi aplicada no Brasi todo.

In [ ]:
#quais colunas sempre True
dataSetSUS.groupby("Ano").count()

072 BCG  ...  009 Haemophilus influenzae b
Ano            ...                              
1994       15  ...                             0
1995       24  ...                            24
1996       26  ...                            26
1997       27  ...                            27
1998       27  ...                            27
1999       27  ...                            27
2000       27  ...                            27
2001       27  ...                            27
2002       27  ...                            27
2003       27  ...                            27
2004       27  ...                             0
2005       27  ...                             0
2006       27  ...                             0
2007       27  ...                             0
2008       27  ...                             0
2009       27  ...                             0
2010       27  ...                             0
2011       27  ...                             0
2012       27  ...                             0
2013       27  ...                             0
2014       27  ...                             0
2016       27  ...                             0
2017       27  ...                             0
2018       27  ...                             0
2019       27  ...                             0

[25 rows x 26 columns]

###**Analisando os dados por ano...**

Link sensacional para quem conhece SQL: [Pandas e SQL](https://pandas.pydata.org/docs/getting_started/comparison/comparison_with_sql.html)

In [ ]:
vacinas_ano_uf = dataSetSUS.melt(id_vars=["Ano","UF"])
vacinas_ano_uf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16475 entries, 0 to 16474
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Ano       16475 non-null  int64 
 1   UF        16475 non-null  object
 2   variable  16475 non-null  object
 3   value     7244 non-null   object
dtypes: int64(1), object(3)
memory usage: 515.0+ KB


Agora vou agrupar por UFs

In [ ]:
lista_uf = dataSetSUS[dataSetSUS["Ano"]==2019]["UF"].to_list()

Loop por UFs para agrupar início e fim de vacinação

In [ ]:
i = 0
while (i<len(lista_uf)):
  vacinas_inicio = vacinas_ano_uf[(vacinas_ano_uf["value"] != 0) & (vacinas_ano_uf["UF"] == lista_uf[i]) & (vacinas_ano_uf["value"].notna())].groupby("variable").min("Ano")
  vacinas_fim = vacinas_ano_uf[(vacinas_ano_uf["value"] != 0) & (vacinas_ano_uf["UF"] == lista_uf[i]) & (vacinas_ano_uf["value"].notna())].groupby("variable").max("Ano")

  vacinas = pd.merge(vacinas_inicio,vacinas_fim, on='variable')
  vacinas["UF"] = lista_uf[i]
  if i == 0:
    vacinas_br = vacinas
  else:
    vacinas_br = pd.concat([vacinas_br,vacinas])
  i = i+1

In [ ]:
vacinas_br.info()

<class 'pandas.core.frame.DataFrame'>
Index: 675 entries, 003 dTpa gestante to 103 Dupla Adulto (dt)
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Ano_x   675 non-null    int64 
 1   Ano_y   675 non-null    int64 
 2   UF      675 non-null    object
dtypes: int64(2), object(1)
memory usage: 41.1+ KB


##Vacina para Sarampo parou no mesmo ano em todo o Brasil: 2003

Mas ainda há o combate, pois a vacina tríplice D1 protege contra esta e outras doenças.

In [ ]:
vacinas_br.loc["018 Sarampo"]

##**Vacinas que são aplicadas desde a década de 90 no Brasil**

Chacando os índices antes de continuar...

In [ ]:
vacinas_br.index

Index(['003 dTpa gestante', '006 Febre Amarela',
       '009 Haemophilus influenzae b', '012 Pneumocócica', '018 Sarampo',
       '021 Tríplice Viral  D1', '053 Meningococo C', '061 Rotavírus Humano',
       '072 BCG', '073 Hepatite B',
       ...
       '093 Poliomielite(1º ref)',
       '094 Dupla adulto e tríplice acelular gestante',
       '095 Tríplice Bacteriana(DTP)(1º ref)', '096 Hepatite A',
       '097 Tetra Viral(SRC+VZ)', '098 Tríplice Viral  D2',
       '099 Hepatite B  em crianças até 30 dias', '100 Poliomielite 4 anos',
       '102 DTP REF (4 e 6 anos)', '103 Dupla Adulto (dt)'],
      dtype='object', name='variable', length=675)

A vacina BCG (que significa Bacilo de Calmette e Guérin) foi criada em 1921 por Léon Calmette e Alphonse Guérin, dando origem ao nome. Ela protege contra as formas graves da doença, como a meningite tuberculosa e a tuberculose miliar (disseminada pelo corpo através da corrente sanguínea.
> A BCG-ID (intradérmica) deve ser administrada ao nascimento ou o mais precocemente possível. Nos prematuros com menos de 36 semanas, administrar a vacina após 1 mês de vida e 2 kg de peso. Administrar uma dose em crianças menores de cinco anos de idade sem cicatriz vacinal

In [ ]:
vacinas_br.loc["072 BCG"]

A hepatite B é causada pelo vírus B (chamado também de HBV). Uma vez dentro do organismo humano, o vírus ataca os hepatócitos (células do fígado) e começa a se multiplicar, levando à inflamação do órgão.

In [ ]:
vacinas_br.loc["073 Hepatite B"]

Mesmo considerada erradicada do Brasil há quase 30 anos, a Poliomielite ainda é uma doença que gera muita preocupação para saúde pública do país.

Uma das principais causas dessa preocupação é o fato da doença ser a causadora da paralisia infantil.

Por isso, anualmente, é importante ressaltar a importância da vacina contra Poliomielite e os meios de prevenção contra o Poliovírus.

In [ ]:
vacinas_br.loc["074 Poliomielite"]

A febre amarela é uma doença infecciosa febril
aguda transmitida por vetores artrópodes e causada
por um vírus do gênero Flavivirus, família Flaviviridae.
Para viajar a países como Angola, por exemplo, a vacinação é obrigatória.
>A doença foi responsável por grande número de
mortes entre o século XVIII e o início do século XX, com
repetidas epidemias nas regiões tropicais da América
do Sul e na África, seguidas por surtos em locais mais
distantes como América do Norte, Caribe e Europa.
A identificação do Aedes aegypti como transmissor
do vírus, em 1900, foi seguida por ações de controle
do vetor que resultaram em significativo declínio da
doença fora das áreas tropicais endêmicas.

In [ ]:
vacinas_br.loc["006 Febre Amarela"]

A Tríplice Bacteriana (DTP) protege a criança de três doenças: Difteria, tétano e coqueluche.

In [ ]:
vacinas_br.loc["075 DTP"]

A tríplice viral D1 começou em 1999, mas não parou mais.
>Trata-se de vacina atenuada, contendo vírus vivos “enfraquecidos” do sarampo, da rubéola e da caxumba; aminoácidos; albumina humana; sulfato de neomicina; sorbitol e gelatina. Contém também traços de proteína do ovo de galinha usado no processo de fabricação da vacina.

No Brasil, uma das vacinas utilizadas na rede pública contém traços de lactoalbumina (proteína do leite de vaca).

Tríplice porque protege contra Sarampo, caxumba e rubéola.

No ano de 2018, o Brasil enfrentou a reintrodução do vírus do sarampo, com a ocorrência de surtos foram confirmados 10.326 casos. Segundo o Ministério da Saúde, até o mês de agosto deste mesmo ano, no estado do Paraná 231,8 mil crianças de um ano a menores de cinco não tinham sido vacinadas, sendo a cobertura vacinal neste período de apenas 60%.

In [ ]:
vacinas_br.loc["021 Tríplice Viral  D1"]

Ano_x  Ano_y                   UF
variable                                                 
021 Tríplice Viral  D1   1999   2019             Rondônia
021 Tríplice Viral  D1   1999   2019                 Acre
021 Tríplice Viral  D1   1999   2019             Amazonas
021 Tríplice Viral  D1   1999   2019              Roraima
021 Tríplice Viral  D1   1999   2019                 Pará
021 Tríplice Viral  D1   1999   2019                Amapá
021 Tríplice Viral  D1   1999   2019            Tocantins
021 Tríplice Viral  D1   1999   2019             Maranhão
021 Tríplice Viral  D1   1999   2019                Piauí
021 Tríplice Viral  D1   1999   2019                Ceará
021 Tríplice Viral  D1   1999   2019  Rio Grande do Norte
021 Tríplice Viral  D1   1999   2019              Paraíba
021 Tríplice Viral  D1   1999   2019           Pernambuco
021 Tríplice Viral  D1   1999   2019              Alagoas
021 Tríplice Viral  D1   1999   2019              Sergipe
021 Tríplice Viral  D1   1999   2019                Bahia
021 Tríplice Viral  D1   1999   2019         Minas Gerais
021 Tríplice Viral  D1   1999   2019       Espírito Santo
021 Tríplice Viral  D1   1999   2019       Rio de Janeiro
021 Tríplice Viral  D1   1999   2019            São Paulo
021 Tríplice Viral  D1   1999   2019               Paraná
021 Tríplice Viral  D1   1999   2019       Santa Catarina
021 Tríplice Viral  D1   1999   2019    Rio Grande do Sul
021 Tríplice Viral  D1   1999   2019   Mato Grosso do Sul
021 Tríplice Viral  D1   1999   2019          Mato Grosso
021 Tríplice Viral  D1   1999   2019                Goiás
021 Tríplice Viral  D1   1999   2019     Distrito Federal

## **Conclusões**

In [ ]:
1. A vacinação pode não ser perfeita, mas tende a ser amplamente disponível no Brasil.
2. Embora a vacina específica contra o sarampo tenha parado de ser distribuída, a tríplice D1 nunca parou. O surto de sarampo em 2018 não foi por falta de vacinação, mas por descuido das pessoas.
3. O início dos registros em 1994 contém vacinação em 15 das 27 UFs, mas não parece ser problema de distribuição, mas talvez de dados (não consegui comprovar a hipótese). Isso porque vários dos estados mais populosos, como RJ e SP, estão entre os que não possuem dados.
4. Muito difícil criar boas visualizações com gráficos convencionais. Pensei em criar um mapa do país com dados de vacinações ou infográficos, mas não usaria o Seaborn para isso...
> Fica para o futuro... confesso que fiquei um pouco frustrado de não achar algo bom para mostrar, por isso mantive o foco na manipulação dos dados.
5. Algumas vacinas estão sempre disponíveis, mas não são alvo de campanhas de vacinação, como a da Febre Amarela, com campanha recente, devido a surto, mas disponível desde 1994.